Test Failure (20180630-1637): SessionCacheTimeoutTest.testRefreshInvalidation

junit.framework.AssertionFailedError: 2018-06-30-23:57:47:725 Servlet call was not successful: ERROR: Caught exception attempting to call test method sessionGet on servlet session.cache.web.SessionCacheTestServlet
java.lang.AssertionError: Was expecting to get testRefreshInvalidation-foo=bar, but instead got a null session.
	at session.cache.web.SessionCacheTestServlet.sessionGet(SessionCacheTestServlet.java:764)
	at componenttest.app.FATServlet.doGet(FATServlet.java:67)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1255)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:743)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:440)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1208)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1005)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:927)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1011)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:414)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:373)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:532)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:466)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:331)
	at com.ibm.ws.http.channel.internal.inbound.HttpICLReadCallback.complete(HttpICLReadCallback.java:70)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:501)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:571)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:926)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1015)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1153)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:785)

	at com.ibm.ws.session.cache.fat.FATSuite.run(FATSuite.java:251)
	at com.ibm.ws.session.cache.fat.SessionCacheApp.invokeServlet(SessionCacheApp.java:29)
	at com.ibm.ws.session.cache.fat.SessionCacheApp.sessionGet(SessionCacheApp.java:49)
	at com.ibm.ws.session.cache.fat.SessionCacheTimeoutTest.testRefreshInvalidation(SessionCacheTimeoutTest.java:165)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:194)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:343)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:168)
	at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:110)

See highlighted time stamps/line below.  In one case it take an extra 3 seconds after the session attribute is read before the servlet call returns.  The test then waits another 3 seconds before trying to access it, but that adds up to 6 seconds which is larger than the 5 second timeout.

[6/30/18 23:57:38:731 UTC] 00000025 SystemOut                                                    O Put entry: testRefreshInvalidation-foo=bar into sessionID=SIr5F-E4uSHnPuJOPIAgc9u
[6/30/18 23:57:41:774 UTC] 00000059 SystemOut                                                    O Got entry: testRefreshInvalidation-foo=bar from sessionID=SIr5F-E4uSHnPuJOPIAgc9u
[6/30/18 23:57:44:637 UTC] 00000059 SystemOut                                                    O <<< END:   sessionGet
[6/30/18 23:57:47:680 UTC] 0000005c SystemOut                                                    O SessionListener1 notified of sessionDestroyed for SIr5F-E4uSHnPuJOPIAgc9u
[6/30/18 23:57:47:691 UTC] 0000005c SystemOut                                                    O ERROR: java.lang.AssertionError: Was expecting to get testRefreshInvalidation-foo=bar, but instead got a null session.

The test case appears to have logic built in for tolerating infrequent timing situations, but needs to be updated for the above.

--- a/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/SessionCacheTimeoutTest.java
+++ b/dev/com.ibm.ws.session.cache_fat/fat/src/com/ibm/ws/session/cache/fat/SessionCacheTimeoutTest.java
@@ -164,9 +164,10 @@ public class SessionCacheTimeoutTest extends FATServletClient {
 
             // Read the session attribute every 3 seconds, looping several times.  Reading the session attribute will
             // prevent the session from becoming invalid after 5 seconds because it refreshes the timer on each access.
-            long start = 0;
+            long start = 0, prevStart = 0;
             try {
                 for (int i = 0; i < refreshes; i++) {
+                    prevStart = start;
                     start = System.nanoTime();
                     TimeUnit.SECONDS.sleep(3);
                     app.sessionGet("testRefreshInvalidation-foo", "bar", session);
@@ -174,7 +175,8 @@ public class SessionCacheTimeoutTest extends FATServletClient {
                 return; // test successful
             } catch (AssertionError e) {
                 long elapsed = System.nanoTime() - start;
-                if (TimeUnit.NANOSECONDS.toMillis(elapsed) > 4500) {
+                if (TimeUnit.NANOSECONDS.toMillis(elapsed) > 4500
+                    || prevStart > 0 && start - prevStart > TimeUnit.SECONDS.toNanos(4)) {
                     Log.info(c, testName.getMethodName(), "Ignoring failure because too much time has elapsed (slow sytem)");
                     continue;
                 } else {

